### PR TITLE
Extracting SubjectKeyId and AuthorityKeyId in case of x509 V3 extensions

### PR DESCRIPTION
**Pull request by tothg001**
Extracting SubjectKeyId and AuthorityKeyId in case of x509 V3 extensions. Updating mbedtls_x509_crt_free function to also free the new dynamic elements (issuer field of AuthorityKeyId).

A few tests are also added which test the feature with a correct certificate and multiple ones with erroneous ASN1 tags.

Signed-off-by: toth92g <toth92g@gmail.com>

## Status
IN DEVELOPMENT

## Requires Backporting
NO

## Migrations
NO

## Todos
- [x] Tests
- [x] Documentation
- [x] Changelog updated

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
```
tests/scripts/all.sh test_memory_buffer_allocator
```
is failing consistently on the CI. The following 5 test cases fail:
```
DTLS fragmenting: proxy MTU, ChachaPoly renego ......................... FAIL
  ! bad client exit code (expected 0, got 1)
  ! outputs saved to o-XXX-660.log
DTLS fragmenting: proxy MTU, AES-GCM renego ............................ FAIL
  ! bad client exit code (expected 0, got 1)
  ! outputs saved to o-XXX-661.log
DTLS fragmenting: proxy MTU, AES-CCM renego ............................ FAIL
  ! bad client exit code (expected 0, got 1)
  ! outputs saved to o-XXX-662.log
DTLS fragmenting: proxy MTU, AES-CBC EtM renego ........................ FAIL
  ! bad client exit code (expected 0, got 1)
  ! outputs saved to o-XXX-663.log
DTLS fragmenting: proxy MTU, AES-CBC non-EtM renego .................... FAIL
  ! bad client exit code (expected 0, got 1)
  ! outputs saved to o-XXX-664.log
```

Here are the logs saved on the CI:
[4140-head-5.zip](https://github.com/ARMmbed/mbedtls/files/5999346/4140-head-5.zip)

File: tests/data_files/Makefile [hunk 1]
```diff
@@ -315,6 +315,15 @@ rsa_pkcs8_2048_public.der: rsa_pkcs8_2048_public.pem
 	$(OPENSSL) rsa -pubin -in $< -outform DER -pubout -out $@
 all_final += rsa_pkcs8_2048_public.der
 
+authorityKeyId_subjectKeyId.crt:
```
Comment (d3zd3z) [inline tests/data_files/Makefile:318]: There is an extraneous space at the end of this line.

File: tests/data_files/Makefile [hunk 2]
```diff
@@ -315,6 +315,15 @@ rsa_pkcs8_2048_public.der: rsa_pkcs8_2048_public.pem
 	$(OPENSSL) rsa -pubin -in $< -outform DER -pubout -out $@
 all_final += rsa_pkcs8_2048_public.der
 
+authorityKeyId_subjectKeyId.crt: 
+	$(OPENSSL) req -x509 -nodes -days 7300 -newkey rsa:2048 -keyout authorityKeyId_subjectKeyId.crt -out authorityKeyId_subjectKeyId.crt -config authorityKeyId_subjectKeyId.conf -extensions 'v3_req'
+# The listed certificates are the copies of authorityKeyId_subjectKeyId.crt with error injections
+# authorityKeyId_subjectKeyId_wrong_SubjectKeyId.crt              The TAG marking the beginning of SubjectKeyId is set to 0x00
+# authorityKeyId_subjectKeyId_wrong_AuthorityKeyId_KeyId.crt      The TAG marking the beginning of AuthorityKeyId field is set to 0x00
+# authorityKeyId_subjectKeyId_wrong_AuthorityKeyId_Sequence.crt   The TAG marking that AuthorityKeyId is a sequence is set to 0x00
+# authorityKeyId_subjectKeyId_wrong_IssuerN.crt                   There are 5 different TAGs based on the x509 doc under AuthorityKeyId(keyId, Dir, Seqence of Dir, serial)
```
Comment (d3zd3z) [inline tests/data_files/Makefile:324]: Extraneous space at the end of this line.

**Review by d3zd3z — COMMENTED** [review]
Two minor whitespace issues.

File: tests/data_files/Makefile (hunk 1 above)
Comment (tothg001) [reply tests/data_files/Makefile]: Thanks for the finding, I have just corrected it

File: tests/data_files/Makefile (hunk 2 above)
Comment (tothg001) [reply tests/data_files/Makefile]: Thanks for the finding, I have just corrected it

**Comment by tothg001** [comment]
Going to be continued in: https://github.com/ARMmbed/mbedtls/pull/4161

Could not solve DCO- missing signoff issue so reset the branch and recommited the changes. After this GIT did not allow to reopen the pull request.

**Comment by tothg001** [comment]
Hello Gilles!

Thank you for the zip. Finally after a whole week of debugging I have found
the problem and corrected it. SSL_client2 and SSL_server2 had too small
heap memory sizes set. 120k was not enough for the certificate objects
after parsing the new fields. Increasing it to 180k solved the issue.
Unfortunately I have again fcked up the signoff of a commit and was not
able to update it via command line. I have hard reset my branch, but could
not reopen the PR again... sorry :(

This is the new PR that is waiting for review:
https://github.com/ARMmbed/mbedtls/pull/4161

Thank you for your help in the last few weeks, I would have failed without
your support.

BR,
Gábor

Gilles Peskine <notifications@github.com> ezt írta (időpont: 2021. febr.
18., Cs, 1:11):

**Pull request closed without merging**